### ISSUE TITLE
Allow project-specific plugins

### ISSUE THREAD

**Issue by rdsubhas** [hashicorp/vagrant#4347]
Overall, I'm getting to #1789 and #3468 (automatic plugin management), but those issues have been closed without any solutions, so here is a shot:

**Step 1:** Add a `--local` flag to `vagrant plugin install`. This will install the plugin into the project's `.vagrant` folder. When you run vagrant up, it should load both global and local plugins.

**Step 2:** Add a Vagrant config option for plugins, so in the Vagrantfile we can say `config.plugins = { ... }`, and then it will internally invoke the above-said local plugin install.

With this, the whole automatic plugin management thing can be taken care using Vagrant's built-in extensibility mechanisms, without having to do any monkey patching to core. How does that sound?

**Comment by habitullence** [comment]
So what's the current way to manage Vagrant plugins?

**Comment by mkuzmin** [comment]
As suggested in the linked issues, I solved this by the next code within `.vagrantplugins` file:
```
required_plugins = %w(
  vagrant-berkshelf
  vagrant-vsphere
  vagrant-guestip
)

needs_restart = false
required_plugins.each do |plugin|
  unless Vagrant.has_plugin? plugin
    system "vagrant plugin install #{plugin}"
    needs_restart = true
  end
end

if needs_restart
  exec "vagrant #{ARGV.join' '}"
end
```

**Comment by habitullence** [comment]
Thanks @mkuzmin, I was thinking about something similar. How come you chose .vagrantplugins as opposed to Vagrantfile? I think I read somewhere that .vagrantplugins is only kept for backwards compatibility and won't be around much longer.

**Comment by mkuzmin** [comment]
no, .vagrantplugins is a new feature introduced in v1.7.
I try to keep my Vagrantfiles short and clear to other people, and move all the complex logic to external files or plugins.

**Comment by r7kamura** [comment]
I solved this problem by creating a tiny plugin [vagrant-multiplug](https://github.com/r7kamura/vagrant-multiplug) that auto-loads plugins configured in Vagrantfile.

**Comment by habitullence** [comment]
@mkuzmin, this seems to contradict @mitchellh, see #5035.

@r7kamura, thanks, saves some keystrokes. Would be good to get this incorporated into Vagrant - what manages the plugin manager.. Bad Watchmen joke :P

**Comment by mkuzmin** [comment]
Here is updated version of `.vagrantplugins` file which supports plugin version locking:
```
required_plugins = {
  'vagrant-berkshelf' => '~>4.1',
  'vagrant-vsphere'   => '~>1.5',
  'vagrant-guestip'   => '0.3.pre1',
  'vagrant-scp'       => '0.5.4',
}

needs_restart = false
required_plugins.each do |name, version|
  unless Vagrant.has_plugin? name, version
    system "vagrant plugin install #{name} --plugin-version=\"#{version}\""
    needs_restart = true
  end
end

if needs_restart
  exec "vagrant #{ARGV.join' '}"
end
```

**Comment by ccope** [comment]
It would be great if you could also specify the source of the plugins being installed via this method. I want to be able to mirror all of our plugins in a local cache to avoid internet flakiness and disappearing gem versions.

**Comment by stantond** [comment]
@chrisroberts this would be fantastic! Is this still planned for one of the 1.9.X releases?

**Comment by paulcdejean** [comment]
I'm relying on the .vagrantplugins file for now, but hopefully this issue will be closed one day.

**Comment by reypm** [comment]
@mkuzmin I know the solution using `.vagrantplugins` is really old but I have two questions:

 - will this work in recent 2.1.2 version?
 - the code inside the `.vagrantplugins` is a plain text? 

@chrisroberts I've seen you tagged this for the new 2.1.3, will this feature be included there?

**Comment by jeffgeorge** [comment]
@reypm I've been using `.vagrantplugins` without issue since late 2016, but it was completely broken by Vagrant 2.1.2. I've confirmed it's still good as of 2.1.1. For reference, yes `.vagrantplugins` is a plaintext file.

Regarding _how_ it has broken, the error I get on 2.1.2 is:
```
Exec error: fork/exec /opt/vagrant/embedded/bin/ruby: argument list too long
[ this repeats endlessly ]
```
I did some searching and this seems related to #8055. There's no root-cause listed in that ticket, but it seemed related to `Vagrant.has_plugin?` choking, with a recommended workaround of "don't install plugins within the scope of a `Vagrantfile`". . . which is unfortunately exactly what we're doing with `.vagrantplugins`.
